File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        
        if(height == null || height.length == 0)
            return 0;
        
        int res = 0, leftMax = 0, rightMax = 0, left = 0, right = height.length - 1;
        
        while(left <= right)
        {
            if(height[left] < height[right])
            {              
                if(height[left] < leftMax)
                    res += leftMax - height[left];
                else
                    leftMax = height[left];
                left++;
            }
            else
            { 
                if(height[right] < rightMax)
                    res += rightMax - height[right];
                else
                    rightMax = height[right];
                right--;
            }
        }
        
        return res;
    }
}

File: 0042-trapping-rain-water/0042-trapping-rain-water.cs
class Solution {
    public int Trap(int[] height) {
        int l = 0;
        int r = 1;
        int res = 0;
        
        while(r < height.Length){
            if(height[r] <= height[l]){
                r ++;
            }else{
                for(int i = l + 1; i < r; i++){
                    res += height[l] - height[i];
                }
                
                l = r;
                r++;
            }
        }
        
        
        
        l = height.Length - 2;
        r = height.Length - 1;
        
        while(l >= 0){
            if(height[l] < height[r]){
                l--;
            }else{
                for(int i = l + 1; i < r; i++){
                    res += height[r] - height[i];
                }
                
                r = l; 
                l--;
            }
        }
        
        return res;
    }
}